Fixed issue with plugins not loaded if multiple plugins are exported in a plugins file

semohr opened the pull request:
It seems like when we simplified the plugin loading, we introduced a regression (#6033).

**This PR introduces a fix:**
We now consider publicly exposed classes listed in a plugin module’s `__all__`.

* If `__all__` is defined, only those members are scanned for `BeetsPlugin` subclasses.
* If multiple plugin classes are exported, we fail fast with a clear error.
* If no `__all__` is defined, we fall back to scanning the full module (preserving backward compatibility).

This makes plugin loading predictable and avoids cases where helper/internal classes are picked up by mistake.

It seems like the `chroma` and `bpsync` plugins exposed an additional `BeetsPlugin` class. Luckily, the [TestImportPlugin](https://github.com/beetbox/beets/blob/b65d773c8e0d309247762ecd50b2fbde336cf816/test/test_plugins.py#L484) test already catches this error if it is raised :+1:

@snejus I think this takes precedence and should be prioritized. Maybe we also do a 2.4.1 release here?


closes #6033 
closes #6023

semohr committed:
Fixed issue with plugins not loaded as multiple plugins are exposed
in the plugins file.

semohr committed:
Replaced filter with list comprehension.

semohr commented:
> Does this not suffice?
> 
> ```diff
>                  and issubclass(obj, BeetsPlugin)
>                  and obj != BeetsPlugin
>                  and not inspect.isabstract(obj)
> +                and obj.__module__ == namespace.__name__
> ```

I was thinking about this too but I don't think this will work. Plugins which are spread across different files and are only exposed in an `__init__.py` file might be an issue. The module does not need to match the namespace in this case.

Haven't tested it tho, just thought about it quickly and disregarded it. I might be wrong here.

semohr commented:
In this specific case it works, but the more general case does not. If the plugin class is not defined directly in the `__init__.py` file but instead imported and re-exported from another module, the `__module__` attribute points to the original file rather than the package namespace.

For example:

```python
# __init__.py
from .foo import MyPlugin

__all__ = ["MyPlugin"]
```

```python
# foo.py
class MyPlugin:
    pass
```

When loaded:

```python
namespace = importlib.import_module("beetsplug.test")
namespace.__name__
>>> 'beetsplug.test'

namespace.MyPlugin.__module__
>>> 'beetsplug.test.foo'
```

This means that relying on `__module__` directly will not always yield the intended package name. While we could work around this with something like a `split` or `startswith`, that still doesn’t fully solve the problem.

A key issue is that there may be multiple plugin classes defined in different files or with inheritance relationships, and only a subset (often just one) should be exposed. For this reason, we should explicitly support and prioritize `__all__`, since it clearly defines the intended public API of the plugin package.

semohr commented:
> I don't think we need to be overly defensive here and account for such a use case, since it adds a fair bit of complexity.

I kinda think we do have to be a bit defensive here. Raising an error here is the proper way to do it. Some developer will run into this issue down the line and we can spare them some time debugging this. Also we will break some plugin otherwise :fearful:

snejus commented:
What about a note in the docs that says 'if your plugin depends on an internal plugin, import the plugin's module' with an example?

semohr commented:
> What about a note in the docs that says 'if your plugin depends on an internal plugin, import the plugin's module' with an example?

Seems a bit hacky to me since we already have the fix here :upside_down_face: Also this is a regression in 2.4.0 so we should restore the original behavior. You that much against the a bit more verbose solution? The only change in the loading is logic just checks the in `__all__` defined

snejus commented:
> You that much against the a bit more verbose solution?

Yes I am, when it fixes an issue which does not exist (once the imports are adjusted as specified in the message above).

semohr commented:
> Sorry for being an ass - I'm just mindful that we have more than enough complexity in the code base, so unless we have no other option, let's keep it simple and stupid.

My problem with the solution is just that the acctual underlying issue is not fixed at all. With the "easier" solution we are just treating symptoms imo.

Maybe a bit of outside perspective could help? @beetbox/maintainers someone else wants to have a look here?

snejus commented:
> Lol the issue still exists it is just not visible anymore. Just sweeped under the carpet. E.g. external plugins could still show the same pattern...

Look, a couple of messages above I changed my mind once you've given me evidence regarding

> I'm doing this in the aisauce plugin 🥲
> 
> Edit: Other plugins I found which do that:
> 
>     * beets-mpd-utils
> 
>     * beets-gogdsync
> 
>     * beets-stylize
> 
>     * beets-importmodifyinfo

---

On a similar note, can you find an external plugin that is now broken because of this issue?

So far, this issue has been limited to two internal plugins, and we have a simple fix for it. In order to consider a more bulletproof/reliable (but more complex) solution, I want to see a real-world example that it's applicable to, rather than a theoretical one.

snejus reviewed (changes requested):
The fix to this regression should _restore_ the previous behavior, rather than requiring the ecosystem to adapt to this issue. This shifts the burden to plugin developers and requires them to adjust their source code.

We have previously been able to load a plugin regardless of whether any other plugin was imported into the same module. We want to continue being able to do so, and we do not want to depend on plugin developers' due diligence to ensure it.

This check should be a good candidate

```py
obj.__module__ == namespace.__path__ or obj.__module__.startswith(f"{namespace.__path__}.")
```

snejus commented:
> also introduce __all__ to make plugin import behavior more explicit and well-defined going forward

Two reasons why we should **not** do this:

1. You specified the first one yourself
> a. A single plugin namespace should not export more than one plugin class (bijective mapping).

Our plugin loading mechanism relies on finding a single plugin class definition (our docs make this pretty clear). Someone attempting to export multiple plugins from the same namespace is misusing the API, thus it's their responsibility to redesign their plugin, once they notice that only the first plugin is loaded.

2. This is purely theoretical - we don't have a real world example to apply this to.

snejus commented:
> I just looked at the dev docs and I dont think we mention that, atleast for me it is not clear 🤷

Feel free to clarify this in the docs then. Though, I don't think there's a need to do so - after all, it seems like no one has encountered this issue in the last 15 years.

I think this is a textbook case of [YAGNI](https://www.geeksforgeeks.org/software-engineering/what-is-yagni-principle-you-arent-gonna-need-it/) violation. And ultimately, the point of this PR is to restore the previous behaviour **for all plugins**, regardless of how many classes they provide.

semohr commented:
> Feel free to clarify this in the docs then. Though, I don't think there's a need to do so - after all, it seems like no one has encountered this issue in the last 15 years.

I will try once again: We could have prevented this in the first place i.e. the regression. It is not primarily about the external plugin developer for me. A simple assertion or a raise would have shown this issue early. It will just prevent people (also us) doing stupid things down the line. Also it aligns expectations with the actual implementation. 

> I think this is a textbook case of [YAGNI](https://www.geeksforgeeks.org/software-engineering/what-is-yagni-principle-you-arent-gonna-need-it/) violation. 

This makes no sense in this context... We are not talking about a new feature here which we need to continuously support. It is just a tiny improvement. You are interpreting this wrong! Also, the 600 open issues paint a different picture. More robustness will help us in the long run.
 
 > And ultimately, the point of this PR is to restore the previous behaviour for all plugins
 
Introducing both fixes as proposed restores the behaviour. 

> , regardless of how many classes they provide.

Allowing multiple plugin classes per namespace would require some changes to the core. While possible it is not what beets internally expects and I don't think it should. This would be an actual example of YAGNI.

JOJ0 commented:
> I disagree! We should have a friendly plugin architecture that allows devs of any level to contribute and write plugins. If a developer tries to write a plugin without fully understanding the underlying architecture, as is often the case, we can’t really blame them if the plugin fails to load or if they decide not to continue contributing. Personally, I would find that frustrating too.
> 
> I don't say we need to check `__all__`, just raising an error in this case will be good too. Just doing anything instead of letting it happen silently. Most importantly if this would have been enforced from the beginning we would have never had this issue in the first place. Or would have catched the issue during the refactor. I just want to make things more maintainable going forward and let us focus on other things by making this more robust.

Guys sometimes a compromise needs to be made in life. I do not see why this solution of @semohr's is not good. Clearly errorying and pointing to usage of __all__ is super clear and plugin dev friendly.

Even though there is not thousands of plugins that run into this it is not bad to clearly raise something to help developers that run into it in the future. This solution is not really much more code than that other solution over there: https://github.com/beetbox/beets/pull/6039 

@wisp3rwind what is your opinion on this? It seems like we need tiebreakers here to be able to move on. Thanks in advance!

wisp3rwind commented:
So, I didn't follow this conversation closely and have only very quickly read through it, so there's definitely a chance that I missed a point or even completely misunderstood something. I don't have a very clear opinion, but a few thoughts:

While this issue hasn't showed up through beets existence before, I do have the impression that the previous implementation (where finding re-"exported" plugin classes in unrelated plugin modules was possible and just happened to work due to how plugin classes were stored in a dict) wasn't well thought out and probably worked rather by chance than conscious design. I think it makes sense to clean this up and have a better documented and more explicit way of detecting/declaring the intended plugin class.

For backwards compatibility, I think we should try to get as close to the old behavior as possible, without requiring changes to plugins. In that sense, Option 1, i.e. the

```py
obj.__module__ == namespace.__path__ or obj.__module__.startswith(f"{namespace.__path__}.")
```

check (which is essentially https://github.com/beetbox/beets/pull/6039 if I didn't miss anything) seems sensible. From the above, it also appears that everybody would be fine with that?

Then, the remaining question is whether we should also add the additional features of Option 2,

- (a) detect the ambiguous case where a plugin defines several candidate classes (e.g. base class + final plugin class) and produce an error in that case
- (b) provide an additional way to disambiguate in that case (e.g. via `__all__`)

Regarding (a), relying on any implicit ordering to determine the class seems brittle at a first glance. It's probably not much of an issue in reality as pointed out above: First, if it happens, it's figured out once by plugin authors when testing their plugin and won't be seen by users. Second, because dict ordering is deterministic these days by preserving insertion order IIRC. Nevertheless, I'm in favor of checking for this, it just doesn't seem like a great "API" to rely on definition order in a plugin module [^2]. Giving a hint to developers here seems like a nice gesture without a big increase in complexity.
However, I think this should produce a warning initially (and load the first class in the dict), which we might escalate to an exception at some point.

Regarding (b), I'm less sure about the necessity, but I'm fine with doing this as well. It probably does not improve plugin development ergonomics by a huge margin (considering the beatport diff given above), but I do like that it is rather explicit, and `__all__` does seem like an obvious way to declare the intended class which I could imagine people try even without checking our docs. (You could argue the reverse here, that it would be unexpected if `__all__` didn't work, given that for standard imports, it does have the desired effect.) Looking at this PR, the complexity of supporting this is also rather low. (Although I'd be happy about a code comment explaining what's going in the various `getattr`s.)

---

So, in summary, I tend to say: Let's merge (something like) @snejus PR quickly[^3], to leave room for any discussion around how to do any further changes, but do take the latter seriously.   
I didn't look at the docs, but it would be great if whatever we do, we make sure that the behavior and how to deal with it is explained clearly. 

[^2]: Although that's generally a thing in Python where modules are executed in definition order upon import, which showed for example in type annotations before they started to be evaluated lazily.
[^3]: It does some additional refactoring, which I didn't check in detail, but at least the importlib conversion seems sensible. In fact, I had prototyped that as well in https://github.com/beetbox/beets/pull/5739, but I let that linger...

JOJ0 commented:
Thanks @wisp3rwind!. I've just approved #6039 and want to ask if someone wants to find a way to remember the "making this future proof parts". Change this PR? Or open a new PR right away? Not sure what's best.

>Then, the remaining question is whether we should also add the additional features of Option 2,

> (a) detect the ambiguous case where a plugin defines several candidate classes (e.g. base class + final plugin class) and produce an error in that case
> (b) provide an additional way to disambiguate in that case (e.g. via __all__)

semohr commented:
>  First, if it happens, it's figured out once by plugin authors when testing their plugin and won't be seen by users... Giving a hint to developers here seems like a nice gesture without a big increase in complexity.

Beyond helping plugin developers, it also gives us a way to verify the intended behavior. While I don’t expect the internal plugin import logic to change often, a recent change here  (#5887)  was the original cause of this issue. Having a check in place, one that we are actually able to test will be a improvement. For me this is quite significant portion of the fix and one if not the main reason for me to push this hard here.

> Regarding (b), I'm less sure about the necessity, ... does seem like an obvious way to declare the intended class which I could imagine people try even without checking our docs.

I agree while it is not necessary, it seems like the obvious way how imports should be handled. As mentioned earlier I'm also fine with dropping this.

semohr committed:
Added a test for multiple plugin classes in a single module.

semohr commented:
> Maybe just continue in this PR, if @semohr is still interested in polishing it? I think it would be interesting to see what it looks like with some of the above suggestions for better backwards compatibility applied.

I just rebased. 

> However, I think this should produce a warning initially (and load the first class in the dict), which we might escalate to an exception at some point. 

Added a deprecation warning for python 3.0.0. We can change it to an exception once we do a version bump.

semohr committed:
Removed __all__ as they are not needed here anymore.
